### DISCUSSION TITLE
Changing the title of wezterm windows

### DISCUSSION THREAD

**Discussion by j-xella** [wezterm/wezterm#5039] [Q&A] [answered]
### What Operating System(s) are you running on?

Windows

### WezTerm version

20240128-202157-1e552d76

### Ask your question!

When I have several wezterm windows open, they all have the same title: "wezterm". Which makes it difficult to switch to the right window quickly, as their tiles may look very similar:

![image](https://github.com/wez/wezterm/assets/8959819/36410c63-b935-4ffc-8b32-957684ce7f11)

Is there a possibility to set the title of wezterm window manually?

Note: I run wezterm with the following settings and usually have one tab only:

``` lua
config.hide_tab_bar_if_only_one_tab = true
config.window_decorations = "INTEGRATED_BUTTONS|RESIZE"
```

**Comment by akthe-at** [comment]
Now, I am not going to claim that I am using these features as @wez intended/appropriately but instead of having multiple instances of wezterm open like in your above screenshot I have been using combination of SwitchWorkspaceRelative, ShowLauncherArgs { flags = "FUZZY|WORKSPACES" }, rename_workspace, and the command palette to create multiple workspaces all in one wezterm "window". 

I have been using that to quickly jump between differently named work areas for myself and it feels clean and organized but I still don't know if its more proper than the multiple open window approach that you are showing above. For all I know they might be equal underneith the hood.

**Comment by kdien** [comment, marked as answer]
You could likely get creative here
```lua
wezterm.on('format-window-title', function()
  local title = 'wezterm'
  -- some logic here
  return title
end)
```

**Reply by j-xella** [reply]
That is exactly what I needed. Here is my title string:

```lua
-- making window titles more distinct
wezterm.on('format-window-title', function()
  local title = '[' .. wezterm.mux.get_active_workspace() .. ']'
  title = title .. ' ' .. wezterm.mux.get_domain():name()
  title = title .. ' - $W'
  -- some logic here
  return title
end)
```

**Comment by yaolifeng0629** [comment]
I want to achieve when I enter a project path, the current window shows me the root directory name of the project I am currently active tab, how to operate?

**Reply by yaolifeng0629** [reply]
The following code does not work either, and there is really no problem to find.

```lua
wezterm.on('format-window-title', function(window, tab)
      local pane = tab.active_pane
      if not pane then
         return nil
      end

      local current_dir = pane.current_working_dir
      if not current_dir then
         return nil
      end

      local path_parts = wezterm.split_string(current_dir, '/')

      local root_path_name = path_parts[1]

      local title = '[' .. wezterm.mux.get_active_workspace() .. ']'
      title = title .. ' ' .. root_path_name
      title = title .. ' - ' .. pane.title

      return title
   end)
```

**Comment by evolsb** [comment]
Yes, and @kdien's sketch is the right shape. Here is the "some logic here" filled in, for the case where you want to name windows **manually** and have the name stick.

The catch worth knowing first: `wezterm cli set-window-title` on its own will not hold. The moment the program in the active tab emits an OSC 0/2 title, WezTerm recomputes the window title from that tab and your name is gone. So the name has to live somewhere durable and be returned from `format-window-title`, which always wins:

```lua
local LABEL_DIR = os.getenv('HOME') .. '/.cache/wezterm-window-labels'

local function read_label(window_id)
  local f = io.open(LABEL_DIR .. '/' .. tostring(window_id), 'r')
  if not f then return nil end
  local v = f:read('*l')
  f:close()
  return (v and #v > 0) and v or nil
end

wezterm.on('format-window-title', function(tab)
  return read_label(tab.window_id) or tab.active_pane.title
end)
```

Bind a key to name the current window:

```lua
{
  key = 'w', mods = 'CMD|SHIFT',
  action = wezterm.action.PromptInputLine {
    description = 'Name this window',
    action = wezterm.action_callback(function(win, _, line)
      if line and #line > 0 then
        local mw = win:mux_window()
        local f = io.open(LABEL_DIR .. '/' .. tostring(mw:window_id()), 'w')
        if f then f:write(line); f:close() end
        mw:set_title(line)
      end
    end),
  },
}
```

A file rather than `wezterm.GLOBAL` because `GLOBAL` only exists inside the GUI process, so a launcher script that just did `wezterm cli spawn --new-window` can also write the name for the window it created.

On the workspaces suggestion: worth flagging that workspaces are **mutually exclusive**, so only one workspace's windows are visible at a time and switching hides the rest. That is great if you want a virtual-desktop model, but it is the wrong tool if (like the screenshot in the original post) you want several named windows visible side by side and want to alt-tab between them. That distinction sent me down a wrong path for a while.

Since you already run `INTEGRATED_BUTTONS|RESIZE`: that also lets you tint the bar per project with a per-window `set_config_overrides { window_frame = { ... } }`, which makes windows even easier to tell apart than the title alone. One caveat there, set **both** `active_titlebar_bg` and `inactive_titlebar_bg` to the same colour, otherwise only the focused window shows its colour and the scheme reads as random rather than as a code.

Full standalone config plus notes: https://gist.github.com/evolsb/35b47be912b53bc706b6b60f2ecff597
